.devcontainer.json

```diff
@@ -1,41 +1,47 @@
 {
-    "image": "mcr.microsoft.com/vscode/devcontainers/javascript-node:18",
-    "containerEnv": {
-      "DATA_WORKSPACE": "/workspaces/data"
+  "image": "mcr.microsoft.com/vscode/devcontainers/javascript-node:18",
+  "containerEnv": {
+    "DATA_WORKSPACE": "/workspaces/data"
+  },
+  "remoteEnv": {
+    "VITE_BUNDLE_HOST": "https://${localEnv:CODESPACE_NAME}-3141.${localEnv:GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN}",
+    "VITE_BUNDLE_SSE": "true"
+  },
+  "onCreateCommand": "rm -rf web && git clone https://github.com/pi-base/web.git && cd web && pnpm install && pnpm run --filter core build",
+  "postAttachCommand": {
+    "viewer": "cd web && pnpm run --filter viewer dev --host --open",
+    "compile": "cd web && pnpm run --filter compile dev",
+    "core": "cd web && pnpm run --filter core dev",
+    // See https://github.com/orgs/community/discussions/4068
+    "publicize": "gh codespace ports visibility 3141:public -c $CODESPACE_NAME"
```

Comment:
I'm a little worried this might be flaky – if there's any race conditions between this running, the compiler server (re)binding to the port, general codespace spin-up things, &c – but it's seeming to work in spot-checking. Feedback _definitely_ appreciated as you use this more.

(I do still want to work on the "fetch on the server side" thing; that should be more reliable anyways. pi-base/web#116)

Reply:
I think this is usable, if as you said, flakey. I haven't looked at the internals, but I can change a file in the Codespace, and get it to refresh in the browser without using the /dev route.

Note that I'm still hitting phantom 404s, e.g. I edit a theorem, and the app doesn't believe the route exists until I navigate to it internally (rather than loading the URL directly).